### PR TITLE
MM-44498: Fix typing UX on channel name input

### PR DESCRIPTION
**Pull request by calebroseland**
#### Summary

- Stabilizes `updatePlaybook` identity 
- Adds `useProxyState` hook to debounce side effects or upstream state-setting
- Debounces channel name input with `useProxyState`

CleanShot 2022-05-27 at 10 32 22

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-44498

#### Checklist

Item | Y / N / NA [^1]
---|---
Telemetry updated? |
Gated by experimental feature flag? |
Unit/E2E tests updated? |

[^1]: Yes / No / Not Applicable.

### CODE REVIEW

File: webapp/src/hooks/general.ts [hunk 1]
```diff
@@ -662,3 +664,47 @@ export const useScrollListener = (el: HTMLElement | null, listener: EventListene
...
+    const [value, setValue] = useState(prop);
+
+    useUpdateEffect(() => {
+        if (!isEqual(value, check.current)) {
+            // check failed; don't destroy pending changes (values set mid-cycle between send/sync)
+            return;
+        }
+        check.current = prop; // sync check
+        setValue(prop);
+    }, [prop]);
```
Comment (calebroseland) [inline webapp/src/hooks/general.ts:694]: This update effect synchronizes the inner value when the prop changes—multi-tab editing, multiple people editing different things on the same playbook, etc.—but only if the check passes. This prevents lost chars and typing glitches without cutting off `value` from outer-updates.

**Review by crspeller — APPROVED** [review]
This is super complicated. I am not sure if I totally understand it. But I also understand why we need it. Hopefully we can remove this with some of the GraphQL performance work we might be doing soon.

**Comment by calebroseland** [comment]
> This is super complicated. I am not sure if I totally understand it. But I also understand why we need it. Hopefully we can remove this with some of the GraphQL performance work we might be doing soon.


> This is super complicated. I am not sure if I totally understand it. But I also understand why we need it. Hopefully we can remove this with some of the GraphQL performance work we might be doing soon.

Very much agreed that this should be temporary—it'd be better to fix the root of the problem in the GraphQL layer, that solution would also scale better to fix other things not behind an edit+save gate like drag&drop, toggles, etc.

**Comment by calebroseland** [comment]
> This is super complicated. I am not sure if I totally understand it. But I also understand why we need it. Hopefully we can remove this with some of the GraphQL performance work we might be doing soon.


> This is super complicated. I am not sure if I totally understand it. But I also understand why we need it. Hopefully we can remove this with some of the GraphQL performance work we might be doing soon.

Very much agreed that this should be temporary—it'd be better to fix the root of the problem in the GraphQL layer, that solution would also scale better to fix other things not behind an edit+save gate like drag&drop, toggles, etc.

**Review by iomodo — APPROVED** [review]

**Pull request merged**